Document `json(field,path)` function (#539)
directus/docs: content/_partials/json-function.md

Reviewer: May be time to add a "Functions" section to the "Connect" links?

Reviewer: This can be disregarded, we will leave it where it is for now.

## The `json()` Function

The `json()` function extracts a specific value from a JSON field and returns it as a separate field in the query response. It is used in the `fields` query parameter alongside regular field names and other field functions.

> > **Note:** the `json()` function is currently only supported for use in the `fields` query parameter.

Reviewer: ```suggestion
The `json(field, path)` function extracts the value at a specified key path within a JSON field and returns it as a separate field in the query response.

::callout{icon="material-symbols:warning-rounded" color="warning"}
**Supported Parameters**
The `json(field, path)` function is currently only supported for use in the `fields` query parameter.
::
```

### Syntax

```
json(field, path)
```

> - **`field`** — the name of a JSON column in the collection (or a relational path to one, see [Relational Queries](#relational-queries)).

Reviewer: ```suggestion
- **`field`** — The field name or [relational field path](#relationa-queries) to the JSON field within the collection.
```

> - **`path`** — a dot-and-bracket notation path to the value you want to extract from within the JSON document.

Reviewer: ```suggestion
- **`path`** — A path using [dot-and-bracket notation](#path-notation) to the JSON property value you want to extract from the document.
```

      }
    }
  ]
}
```

#### One-to-Many (O2M)

For O2M relations, each related item returns its own extracted value. The response is an array of objects, each containing the extracted key.

```http
GET /items/articles/1?fields=id,json(comments.data, type)
```

```json
{
  "data": {
    "id": 1,
    "comments": [
      { "data_type_json": "review" },
      { "data_type_json": "feedback" },
      { "data_type_json": "question" }
    ]
  }
}
```

#### Any-to-One (A2O / Many-to-Any)

For A2O relations, use the standard Directus collection scope syntax inside the first argument:

```
json(relation.item:collection_name.json_field, path)
```

```http
GET /items/shapes/1?fields=id,json(children.item:circles.metadata, color)
```

### Relational Depth Limit

Directus enforces a maximum relational depth (`MAX_RELATIONAL_DEPTH`, default `10`). For `json()`, only the **first argument** (the field path) counts toward this limit — the JSON path in the second argument is excluded.

```
json(category_id.metadata, a.b.c.d.e)
```
This has a relational depth of **2** (`category_id` + `metadata`), regardless of how many segments are in the JSON path `a.b.c.d.e`.

> Exceeding the relational depth returns a `400 Invalid query. Max relational depth exceeded.` error.

Reviewer: ```suggestion
Exceeding the relational depth  will return an error.
```

### JSON Path Depth Limit

The JSON path (second argument) also has its own configurable limit:

```
MAX_JSON_QUERY_DEPTH=10  # default
```

Depth is counted by the number of `.` and `[` characters in the normalized path. A path like `a.b.c.d.e.f.g.h.i.j` has depth 10 and is allowed by default; adding one more segment exceeds the limit.

> Exceeding this limit returns an error.

Reviewer: Let's keep the format for this section inline with `Relational Depth Limit`

### Unsupported Path Expressions

The following path syntaxes are **not supported** and will return a `400` error:

| Expression | Example |
|---|---|---|
| Empty brackets (wildcard) | `items[]` |
| `[*]` wildcard | `items[*].name` |
| `*` glob | `items.*` |
| JSONPath predicates | `items[?(@.price > 10)]` |
| `@` current node | `@.name` |
| `$` root | `$.name` |

### Object Keys with Special Characters

The `json()` path syntax uses `.` as a separator between key segments. There is no escape mechanism for object keys that themselves contain dots, spaces, or other special characters. For example, if your JSON has a key `"first.name"`, there is no way to express that in the path — `json(data, first.name)` would be interpreted as nested access to key `first`, then key `name`.

Similarly, because MySQL and MariaDB path conversion uses dot-notation (`$.key.subkey`), keys containing characters that are special in that context (e.g., spaces) may not be reachable. PostgreSQL's parameterized `->?` approach is more permissive for unusual key names, but the input path format still does not provide an escaping mechanism.

> ### Database-Specific Exceptions

Reviewer: ```suggestion
### Database-Specific Caveats
```

**SQLite**

> - SQLite can return `0`/`1` isntead of `boolean` values.

Reviewer: ```suggestion
- Can return `0`/`1` isntead of `boolean` values.
```

**MSSQL**

> - `JSON_VALUE` in MSSQL always returns scalar values as **strings (`NVARCHAR`)**, even when the original JSON value is a number or boolean. For example, a JSON integer `42` will be returned as the string `"42"`. Your application should perform type coercion as needed.

Reviewer: ```suggestion
- Will always returns scalar values as **strings (`NVARCHAR`)**, even when the original JSON value is a number or boolean. For example, a JSON integer `42` will be returned as the string `"42"`. Your application should perform type coercion as needed.
```

**Oracle**

> - Similar to MSSQL, `JSON_VALUE` in Oracle returns scalar values as **strings**, regardless of the original JSON type (number, boolean, etc.). A JSON number `3.14` will be returned as `"3.14"`.

Reviewer: ```suggestion
- Similar to MSSQL will also return scalar values as **strings**, regardless of the original JSON type (number, boolean, etc.). A JSON number `3.14` will be returned as `"3.14"`.
```

        "metadata_icon_json": "laptop"
      }
    }
  ]
}
```

#### One-to-Many (O2M)

For O2M relations, each related item returns its own extracted value. The response is an array of objects, each containing the extracted key.

```http
GET /items/articles/1?fields=id,json(comments.data, type)
```

```json
{
  "data": {
    "id": 1,
    "comments": [
      { "data_type_json": "review" },
      { "data_type_json": "feedback" },
      { "data_type_json": "question" }
    ]
  }
}
```

#### Many-to-Any (M2A)

For M2A relations, use the standard Directus collection scope syntax inside the first argument:

```
json(relation.item:collection_name.json_field, path)
```

```http
GET /items/shapes/1?fields=id,json(children.item:circles.metadata, color)
```

### Relational Depth Limit

Directus enforces a maximum relational depth (`MAX_RELATIONAL_DEPTH`, default `10`). For `json()`, only the **first argument** (the field path) counts toward this limit — the JSON path in the second argument is excluded.

```
json(category_id.metadata, a.b.c.d.e)
```
This has a relational depth of **2** (`category_id` + `metadata`), regardless of how many segments are in the JSON path `a.b.c.d.e`.

> Exceeding the relational depth  will return an error.

Reviewer: ```suggestion
Exceeding the relational depth will return an error.
```
